src/gt4py/next/iterator/type_inference.py

```diff
@@ -418,26 +420,30 @@ def visit_Sym(self, node: ir.Sym, **kwargs) -> Type:
         return TypeVar.fresh()
 
     def visit_SymRef(self, node: ir.SymRef, *, symtable, **kwargs) -> Type:
-        if node.id in BUILTIN_TYPES:
-            return freshen(BUILTIN_TYPES[node.id])
-        if node.id in (
-            "make_tuple",
-            "tuple_get",
-            "shift",
-            "cartesian_domain",
-            "unstructured_domain",
-            "cast_",
-        ):
-            raise TypeError(
-                f"Builtin '{node.id}' is only supported as applied/called function by the type checker"
-            )
-        if node.id in symtable:
-            res = self.collected_types[id(symtable[node.id])]
+        if node.id in ir.BUILTINS:
+            if node.id in BUILTIN_TYPES:
+                return freshen(BUILTIN_TYPES[node.id])
+            elif node.id in (
+                "make_tuple",
+                "tuple_get",
+                "shift",
+                "cartesian_domain",
+                "unstructured_domain",
+                "cast_",
+                ir.TYPEBUILTINS,
```

Comment:
Should `TYPEBUILTINS` be here?

```
def foo():
    return int64
cast_(expr, foo())
```

Reply:
Cast is currently part of the grammar as we otherwise need the type of type thing to propagate the information. Hence the above does not work. The first thing when implementing such functionality would be to remove `ir.TYPEBUILTINS` here and instead return the type of type.

Reply:
Let's add a comment